### PR TITLE
Тышкевич Артем. Задача 1. Вариант 6. Нахождение числа нарушений упорядоченности соседних элементов вектора.

### PR DESCRIPTION
**Pull request by artich02**
В последовательной реализации вектор инициализируется, затем выполняется проход по массиву с поэлементным сравнением. При этом вектор нулевого размера воспринимается, как ошибка на этапе валидации. Если текущий элемент больше следующего, счётчик нарушений увеличивается.

В параллельной версии (MPI) вектор делится на части, и каждый процесс обрабатывает свою часть в отдельном буфере. Root-процесс распределяет данные, и каждый non-root процесс подсчитывает нарушения локально. Для корректного учёта нарушений на границах подмассивов используется смещение данных на один элемент. По завершении вычислений результаты собираются и суммируются в root-процессе.

### CODE REVIEW

File: tasks/mpi/tyshkevich_a_num_of_orderly_violations/func_tests/main.cpp [hunk 1]
```diff
@@ -0,0 +1,162 @@
+#include <gtest/gtest.h>
+
+#include <boost/mpi/communicator.hpp>
+#include <boost/mpi/environment.hpp>
+#include <string>
+#include <vector>
+
+#include "mpi/tyshkevich_a_num_of_orderly_violations/include/ops_mpi.hpp"
```
Comment (ovVrLFg8ks) [inline tasks/mpi/tyshkevich_a_num_of_orderly_violations/func_tests/main.cpp:8]: Add tests for zero-sized input

**Review by ovVrLFg8ks — CHANGES_REQUESTED** [review]

File: tasks/mpi/tyshkevich_a_num_of_orderly_violations/func_tests/main.cpp (hunk 1 above)
Comment (ovVrLFg8ks) [inline tasks/mpi/tyshkevich_a_num_of_orderly_violations/func_tests/main.cpp:8]: Add tests for zero-sized input

**Review by ovVrLFg8ks — CHANGES_REQUESTED** [review]

File: tasks/mpi/tyshkevich_a_num_of_orderly_violations/func_tests/main.cpp (hunk 1 above)
Comment (artich02) [reply tasks/mpi/tyshkevich_a_num_of_orderly_violations/func_tests/main.cpp:8]: added test

**Review by ovVrLFg8ks — APPROVED** [review]

**Comment by artich02** [comment]
> I've noticed several files have unnecessary or indirect include directives, which obscure dependencies. Please review all include directives in the modified files to ensure each file includes only the headers it actually needs.

removed the unnecessary

**Review by andreychh — APPROVED** [review]

File: tasks/mpi/tyshkevich_a_num_of_orderly_violations/func_tests/main.cpp [hunk 2]
```diff
@@ -0,0 +1,210 @@
+#include <gtest/gtest.h>
+
+#include <boost/mpi/communicator.hpp>
+#include <boost/mpi/environment.hpp>
+#include <string>
+#include <vector>
+
+#include "mpi/tyshkevich_a_num_of_orderly_violations/include/ops_mpi.hpp"
+
+std::string VecToStrTY(std::vector<int> v) {
```
Comment (aobolensk) [inline tasks/mpi/tyshkevich_a_num_of_orderly_violations/func_tests/main.cpp:10]: Do we need this copied by value?

File: tasks/seq/tyshkevich_a_num_of_orderly_violations/perf_tests/main.cpp [hunk 3]
```diff
@@ -0,0 +1,82 @@
+#include <gtest/gtest.h>
+
+#include <vector>
+
+#include "core/perf/include/perf.hpp"
+#include "seq/tyshkevich_a_num_of_orderly_violations/include/ops_seq.hpp"
+
+TEST(tyshkevich_a_num_of_orderly_violations_seq_perftest, test_pipeline_run) {
+  int size = 999;
```
Comment (aobolensk) [inline tasks/seq/tyshkevich_a_num_of_orderly_violations/perf_tests/main.cpp:9]: Too small size for perf test

**Review by aobolensk — CHANGES_REQUESTED** [review]

**Comment by artich02** [comment]
@aobolensk I fixed the code based on your comments. But it was not my fault that macos-clang-build crashed. This can be seen from the last two commits.

File: tasks/mpi/tyshkevich_a_num_of_orderly_violations/src/ops_mpi.cpp [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+  }
+  return true;
+}
+
+bool tyshkevich_a_num_of_orderly_violations_mpi::TestMPITaskParallel::run() {
+  internal_order_test();
+
+  if (size == 0) {
+    return true;
+  }
```
Comment (allnes) [inline tasks/mpi/tyshkevich_a_num_of_orderly_violations/src/ops_mpi.cpp:87]: please check it in validation() function

**Review by allnes — CHANGES_REQUESTED** [review]

File: tasks/mpi/tyshkevich_a_num_of_orderly_violations/src/ops_mpi.cpp (hunk 4 above)
Comment (artich02) [reply tasks/mpi/tyshkevich_a_num_of_orderly_violations/src/ops_mpi.cpp:87]: @allnes I can make inputs_count > 0, but I will be forced to remove the test for zero size?

Comment (allnes) [reply tasks/mpi/tyshkevich_a_num_of_orderly_violations/src/ops_mpi.cpp:87]: please remove it, because yours task for size = 0 is not relevant

**Review by allnes — APPROVED** [review]

File: tasks/seq/tyshkevich_a_num_of_orderly_violations/func_tests/main.cpp [hunk 5]
```diff
@@ -0,0 +1,77 @@
+#include <gtest/gtest.h>
+
+#include <vector>
+
+#include "seq/tyshkevich_a_num_of_orderly_violations/include/ops_seq.hpp"
+
+TEST(tyshkevich_a_num_of_orderly_violations_seq_ftest, Test_10) {
```
Comment (aobolensk) [inline tasks/seq/tyshkevich_a_num_of_orderly_violations/func_tests/main.cpp:7]: Please add some more tests from MPI version here as well

File: tasks/mpi/tyshkevich_a_num_of_orderly_violations/src/ops_mpi.cpp [hunk 6]
```diff
@@ -0,0 +1,134 @@
+#include "mpi/tyshkevich_a_num_of_orderly_violations/include/ops_mpi.hpp"
+
+#include <random>
+#include <vector>
+
+using namespace std::chrono_literals;
+
+std::vector<int> tyshkevich_a_num_of_orderly_violations_mpi::getRandomVector(int sz) {
```
Comment (aobolensk) [inline tasks/mpi/tyshkevich_a_num_of_orderly_violations/src/ops_mpi.cpp:8]: Please, move to the test module

**Review by aobolensk — CHANGES_REQUESTED** [review]

File: tasks/mpi/tyshkevich_a_num_of_orderly_violations/src/ops_mpi.cpp (hunk 6 above)
Comment (aobolensk) [reply tasks/mpi/tyshkevich_a_num_of_orderly_violations/src/ops_mpi.cpp]: Exclude getRandomVector declarations from header as well

Comment (artich02) [reply tasks/mpi/tyshkevich_a_num_of_orderly_violations/src/ops_mpi.cpp]: @aobolensk Done
